### PR TITLE
Runtime BC control for both cc and fc variables

### PR DESCRIPTION
**Pull request by chongchonghe**
### Description
Now, we define boundary conditions with a single runtime parameter. This overrides `geometry.is_periodic =  1    1    1`. No need for `BCs_cc` in the problem creator. 
```
quokka.bc = periodic periodic reflecting
```

### Checklist
_Before this pull request can be reviewed, all of these tasks should be completed. Denote completed tasks with an `x` inside the square brackets `[ ]` in the Markdown source below:_
- [ ] I have added a description (see above).
- [ ] I have added a link to any related issues (if applicable; see above).
- [ ] I have read the [Contributing Guide](https://github.com/quokka-astro/quokka/blob/development/CONTRIBUTING.md).
- [ ] I have added tests for any new physics that this PR adds to the code.
- [ ] *(For quokka-astro org members)* I have manually triggered the GPU tests with the magic comment `/azp run`.

### CODE REVIEW

**Comment by chongchonghe** [comment]
@BenWibking  I'll wait for your comments before refactoring all the problems to use this new design, which requires some python scripting.

**Comment by BenWibking** [comment]
This is a nice clean-up. Can it also set `BCs_fc`?

**Comment by chongchonghe** [comment]
> This is a nice clean-up. Can it also set `BCs_fc`?

Not yet, but it's trivial to extend it to support that. Then, there will be a `quokka.BC_fc`

**Comment by BenWibking** [comment]
> > This is a nice clean-up. Can it also set `BCs_fc`?
> 
> Not yet, but it's trivial to extend it to support that. Then, there will be a `quokka.BC_fc`

I think it would be better to set both with the same parameter. Is there a case where you think that wouldn't be possible?

**Comment by chongchonghe** [comment]
> > > This is a nice clean-up. Can it also set `BCs_fc`?
> > 
> > 
> > Not yet, but it's trivial to extend it to support that. Then, there will be a `quokka.BC_fc`
> 
> I think it would be better to set both with the same parameter. Is there a case where you think that wouldn't be possible?

That will be even simpler. I don't know if MHD B.C. will always be the same as hydro. If yes, then we should use the same parameter. 

What about reflecting BC for fc variables? All current MHD tests use periodic or Dirichlet BC, but I think we want to support reflecting BC as well, right? Do we have to handle odd/even reflection, or is it always even reflection since we only have one variable (magnetic field) for each dimension?

**Comment by BenWibking** [comment]
> > > > This is a nice clean-up. Can it also set `BCs_fc`?
> > > 
> > > 
> > > Not yet, but it's trivial to extend it to support that. Then, there will be a `quokka.BC_fc`
> > 
> > 
> > I think it would be better to set both with the same parameter. Is there a case where you think that wouldn't be possible?
> 
> That will be even simpler. I don't know if MHD B.C. will always be the same as hydro. If yes, then we should use the same parameter.
> 
> What about reflecting BC for fc variables? All current MHD tests use periodic or Dirichlet BC, but I think we want to support reflecting BC as well, right? Do we have to handle odd/even reflection, or is it always even reflection since we only have one variable (magnetic field) for each dimension?

You can do reflecting BCs for magnetic fields, but you have to handle the parity in a special way. The normal component should be even, and the transverse components should be odd.

**Comment by chongchonghe** [comment]
> You can do reflecting BCs for magnetic fields, but you have to handle the parity in a special way. The normal component should be even, and the transverse components should be odd.

@BenWibking  I'll leave this for a future PR. In this PR, I assume 'reflecting' for cc = 'reflect_even' for fc, just like what you did in all the MHD particle tests.

File: src/simulation.hpp [hunk 1]
```diff
@@ -220,13 +221,34 @@ template <typename problem_t> class AMRSimulation : public amrex::AmrCore
...
+		amrex::ParmParse const pp_quokka("quokka");
+		amrex::Vector<quokka::BCType::mathematicalBndryTypes> bc_type;
+		if (pp_quokka.queryarr("bc", bc_type) != 0) {
+			// Parse BCs
+			AMREX_ALWAYS_ASSERT_WITH_MESSAGE(bc_type.size() == 3, "quokka.bc must have 3 components");
+
+			BCs_cc_ = quokka::BC_cc<problem_t>(bc_type[0], bc_type[1], bc_type[2]);
+			BCs_fc_ = quokka::BC_fc<problem_t>(bc_type[0], bc_type[1], bc_type[2]);
+		} else {
+			amrex::Abort("quokka.bc must be specified in the input file.");
```
[the remark this answers is not in the document]
Comment (chongchonghe) [reply src/simulation.hpp:248]: Done.

**Comment by BenWibking** [comment]
> > You can do reflecting BCs for magnetic fields, but you have to handle the parity in a special way. The normal component should be even, and the transverse components should be odd.
> 
> @BenWibking I'll leave this for a future PR. In this PR, I assume 'reflecting' for cc = 'reflect_even' for fc, just like what you did in all the MHD particle tests.

@chongchonghe Ok, that's fine. Can you create an issue for this so we don't forget about it?

Also, can you edit the MHD documentation page to add a subsection on boundary conditions?

**Comment by chongchonghe** [comment]
> Also, can you edit the MHD documentation page to add a subsection on boundary conditions?

Done.

**Comment by chongchonghe** [comment]
@BenWibking Ready for review.

**Comment by chongchonghe** [comment]
@BenWibking This is ready for review.

**Comment by chongchonghe** [comment]
@BenWibking This is ready.

File: docs/markdown/mhd_module.md [hunk 2]
```diff
@@ -126,4 +126,4 @@ more detail in [Runtime parameters](parameters.md):
   (default `Balsara2025`).
 - `artificial_viscosity_k` – optional scalar viscosity coefficient that adds a
   diffusive flux to the momentum equations and can damp post-shock oscillations.
-
+- `quokka.bc` – (required) choose `periodic` or `reflecting` for the boundary conditions. For reflecting boundaries, we use `amrex::BCType::reflect_even` for all magnetic field components. Support for properly reflecting magnetic field boundaries will be added in the future.
```
Comment (BenWibking) [inline docs/markdown/mhd_module.md]: Can you add a line to the "runtime parameters" documentation page as well? We should document that it can be `periodic`, `reflecting`, `foextrap`, or `ext_dir`.

File: inputs/BinaryOrbitAMR.in [hunk 3]
```diff
@@ -3,7 +3,7 @@
 # *****************************************************************
 geometry.prob_lo     =  -2.5e13 -2.5e13 -2.5e13
 geometry.prob_hi     =   2.5e13  2.5e13  2.5e13
```
Comment (markkrumholz) [inline inputs/BinaryOrbitAMR.in:5]: Not sure if you meant, but looks like you have changed things here from periodic in x only to periodic in both x and y.

File: inputs/radhydro_shell_regression.in [hunk 4]
```diff
@@ -3,7 +3,7 @@
 # *****************************************************************
 geometry.prob_lo     =  0.        0.        0.
 geometry.prob_hi     =  3.086e19  3.086e19  3.086e19
-geometry.is_periodic =  0   0   0
```
Comment (markkrumholz) [inline inputs/radhydro_shell_regression.in:6]: Here you have changed from all non-periodic to all periodic.

File: src/main.cpp [hunk 5]
```diff
@@ -17,35 +17,52 @@
 
 auto main(int argc, char **argv) -> int
 {
-	// Initialization (copied from ExaWind)
+	/// Initialization (copied from ExaWind)
```
Comment (markkrumholz) [inline src/main.cpp:20]: Looks like extra comment markers got added here somehow. Not harmful, but should probably be fixed.

File: src/main.cpp [hunk 6]
```diff
@@ -17,35 +17,52 @@
...
 
 	amrex::Initialize(argc, argv, true, MPI_COMM_WORLD, []() {
 		amrex::ParmParse pp("amrex");
-		// Set GPU memory handling defaults:
-		// since performance is terrible if we have to swap pages between device and
-		// host memory due to exceeding the size of device memory, we crash the code
-		// if this happens, allowing the user to restart with more nodes.
+		/// Set GPU memory handling defaults:
```
Comment (markkrumholz) [inline src/main.cpp:24]: More extra comment characters

File: src/main.cpp [hunk 7]
```diff
@@ -17,35 +17,52 @@
...
+		/// since performance is terrible if we have to swap pages between device and
+		/// host memory due to exceeding the size of device memory, we crash the code
+		/// if this happens, allowing the user to restart with more nodes.
 		if (!pp.contains("abort_on_out_of_gpu_memory")) {
 			pp.add("abort_on_out_of_gpu_memory", 1);
 		}
 
-		// disables managed memory
```
Comment (markkrumholz) [inline src/main.cpp:32]: Extra comment characters again

File: src/main.cpp [hunk 8]
```diff
@@ -17,35 +17,52 @@
...
-		//  transport and leads to *extremely poor* GPU-aware MPI performance.
+		/// disables managed memory
+		/// for single-GPU runs, the overhead is completely negligible.
+		/// HOWEVER, for multi-GPU runs, using managed memory disables the cuda_ipc
+		/// transport and leads to *extremely poor* GPU-aware MPI performance.
 		if (!pp.contains("the_arena_is_managed")) {
 			pp.add("the_arena_is_managed", 0);
 		}
 
-		// use GPU-aware MPI
```
Comment (markkrumholz) [inline src/main.cpp:40]: Here too

**Review by markkrumholz — COMMENTED** [review]
Just left some minor cleanup comments.

File: inputs/BinaryOrbitAMR.in (hunk 3 above)
Comment (chongchonghe) [reply inputs/BinaryOrbitAMR.in:5]: I changed it while running testing manually and didn't change back. This doesn't matter. This test works equally well with arbitrary combinations of periodic and open BCs.

Comment (chongchonghe) [reply inputs/BinaryOrbitAMR.in:5]: I changed it while running testing manually and didn't change back. This doesn't matter. This test works equally well with arbitrary combinations of periodic and open BCs.

File: src/main.cpp (hunk 5 above)
Comment (chongchonghe) [reply src/main.cpp:20]: This is intentional. SonarCloud thinks '//' is 'commented out code' and '///' is inline comments, so this is to suppress 'remove commented out code' warnings.

File: inputs/radhydro_shell_regression.in (hunk 4 above)
Comment (markkrumholz) [reply inputs/radhydro_shell_regression.in:6]: This one I think is a real bug -- the radhydro shell test should be using reflecting BC's.

Comment (chongchonghe) [reply inputs/radhydro_shell_regression.in:6]: I just fixed it. Indeed it was a mistake. Actually, in the current development branch, all the radhyro_shell_xxx.in files have wrong values of `geometry.is_periodic` because they should be all '0's. I have changed that.

**Review by markkrumholz — APPROVED** [review]

File: inputs/radhydro_shell_regression.in (hunk 4 above)
Comment (markkrumholz) [reply inputs/radhydro_shell_regression.in:6]: Good catch. Approved.

**Comment by markkrumholz** [comment]
Looks like this will need to be updated to the base branch to pass the rocm-diskgalaxy-container test. Also, there's still a SonarCloud failure that will block merging.

**Comment by chongchonghe** [comment]
> Also, there's still a SonarCloud failure that will block merging.

The SonarCloud warnings are optional and we can ignore it. I think unless we can control the rules of SonarCloud, we should not make it mandatory. We had discussions in #1529
